### PR TITLE
Authorisation helper

### PR DESCRIPTION
**Pull request by Pete-Brown**
Adds methods to generate authorization url, make token requests and make token refresh requests

### CODE REVIEW

File: lib/sageone_sdk/authentication.rb [hunk 1]
```diff
@@ -4,5 +4,39 @@ module Authentication
     def token_authenticated?
       !!@access_token
     end
+
+    def authorization_url
+      "https://www.sageone.com/oauth2/auth?response_type=code&client_id=#{@client_id}&redirect_uri=#{redirect_uri}&scope=full_access"
```
Comment (adamgeorgeson) [inline lib/sageone_sdk/authentication.rb]: Is this url correct? I'd have expected `app.sageone.com` opposed to `www.sageone.com`

Comment (Pete-Brown) [inline lib/sageone_sdk/authentication.rb]: It used to be, however the API documentation now lists it as www.sageone.com/oauth2/auth

https://developers.sageone.com/docs/en/v1#authentication-request_authorization_code

Comment (chrisbarber86) [inline lib/sageone_sdk/authentication.rb]: Yep www.sageone.com is correct. It then proxies through to the correct location

File: lib/sageone_sdk/authentication.rb [hunk 2]
```diff
@@ -4,5 +4,39 @@ module Authentication
...
+    end
+
+    def request_access_token (code)
```
Comment (chrisbarber86) [inline lib/sageone_sdk/authentication.rb:18]: Documentation needed on these methods.

File: lib/sageone_sdk/authentication.rb [hunk 3]
```diff
@@ -4,5 +4,39 @@ module Authentication
...
+      token_request("refresh_token")
+    end
+
+    def token_request(type, options = {})
+      request = Faraday.new() do |builder|
+        builder.request :url_encoded
+        builder.response :json, :content_type => /\bjson$/
+        builder.adapter Faraday.default_adapter
+      end
+      if type == "access_token"
```
Comment (chrisbarber86) [inline lib/sageone_sdk/authentication.rb:39]: could we create payload here as

``` ruby
payload = { client_id: @client_id, client_secret: @client_secret }
```

then in the if branches do:

``` ruby
payload = payload.merge({
  code: options[:code],
  grant_type: 'authorization_code',
  redirect_uri: @redirect_uri
})
```

or 

``` ruby
payload = payload.merge({
  refresh_token: @refresh_token,
  grant_type: 'refresh_token'
})
```

File: lib/sageone_sdk/authentication.rb [hunk 4]
```diff
@@ -4,5 +4,39 @@ module Authentication
...
+                      code: options[:code],
+                      grant_type: 'authorization_code',
+                      redirect_uri: @redirect_uri}
+      elsif type == "refresh_token"
+        payload = {client_id: @client_id,
+                      client_secret: @client_secret,
+                      refresh_token: @refresh_token,
+                      grant_type: 'refresh_token'}
+      end
+      json_response = request.post 'https://api.sageone.com/oauth2/token', payload
```
Comment (chrisbarber86) [inline lib/sageone_sdk/authentication.rb]: This needs to use `API_ENDPOINT`

``` ruby
json_response = request.post "#{API_ENDPOINT}/oauth2/token", payload
```

File: README.md [hunk 5]
```diff
@@ -22,7 +22,40 @@ Or install it yourself as:
 
 ## Usage
 
-To create a `SageoneSdk::Client`, you need to provide your access_token and signing_secret:
+# Authorization
```
Comment (chrisbarber86) [inline README.md]: If these headings should be nested under `## Usage`, they need to be h3's : `### Authorization`

File: lib/sageone_sdk/configurable.rb [hunk 6]
```diff
@@ -1,7 +1,15 @@
 module SageoneSdk
   # Configurable
   module Configurable
-    attr_accessor :access_token, :connection_options, :default_media_type, :user_agent, :signing_secret
+    attr_accessor :access_token
```
Comment (chrisbarber86) [inline lib/sageone_sdk/configurable.rb]: these can stay on the one line:

``` ruby
attr_accessor :access_token, :connection_options, :default_media_type, :user_agent,
              :signing_secret, :refresh_token, :redirect_uri, :client_id, :client_secret
```

**Comment by chrisbarber86** [comment]
:+1: Looks Good

**Pull request merged**